p2p/conn: channel id is not checked for overflow

<!--
Please fill in as much of the template below as you can.

Be ready for followup questions, and please respond in a timely
manner. We might ask you to provide additional logs and data (tendermint & app).
-->

**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source):

4e06dfef8c60680850613e5f337de49f71dab4f0

**What happened**:

This test should run successfully, but it's failed. The channel id `1025` must be rejected.

```
func TestMConnectionChannelOverflow(t *testing.T) {
	chOnErr := make(chan struct{})
	chOnRcv := make(chan struct{})

	mconnClient, mconnServer := newClientAndServerConnsForReadErrors(t, chOnErr)
	t.Cleanup(stopAll(t, mconnClient, mconnServer))

	mconnServer.onReceive = func(chID byte, msgBytes []byte) {
		chOnRcv <- struct{}{}
	}

	client := mconnClient.conn
	protoWriter := protoio.NewDelimitedWriter(client)

	var packet = tmp2p.PacketMsg{
		ChannelID: 0x01,
		EOF:       true,
		Data:      []byte(`42`),
	}
	_, err := protoWriter.WriteMsg(mustWrapPacket(&packet))
	require.NoError(t, err)
	assert.True(t, expectSend(chOnRcv))

	packet.ChannelID = int32(1025)
	_, err = protoWriter.WriteMsg(mustWrapPacket(&packet))
	require.NoError(t, err)
	assert.False(t, expectSend(chOnRcv))

}
```

**What you expected to happen**:

Test passed.

**Have you tried the latest version**:

yes

**How to reproduce it** (as minimally and precisely as possible):

See inlined code above, run the test under `p2p/conn`.

**Logs (paste a small part showing an error (< 10 lines) or link a pastebin, gist, etc. containing more of the log file)**:

```
=== RUN   TestMConnectionChannelOverflow
I[2021-06-01|16:33:27.281] Starting MConnection service                 module=client impl=MConn{pipe}
I[2021-06-01|16:33:27.281] Starting MConnection service                 module=server impl=MConn{pipe}
D[2021-06-01|16:33:27.281] Read PacketMsg                               module=server conn=MConn{pipe} packet="{ChannelID:1 EOF:true Data:[52 50]}"
D[2021-06-01|16:33:27.281] Received bytes                               module=server chID=1 msgBytes=3432
D[2021-06-01|16:33:27.281] Read PacketMsg                               module=server conn=MConn{pipe} packet="{ChannelID:1025 EOF:true Data:[52 50]}"
D[2021-06-01|16:33:27.281] Received bytes                               module=server chID=1025 msgBytes=3432
    connection_test.go:582:
        	Error Trace:	connection_test.go:582
        	Error:      	Should be false
        	Test:       	TestMConnectionChannelOverflow
I[2021-06-01|16:33:27.282] Stopping MConnection service                 module=client impl=MConn{pipe}
I[2021-06-01|16:33:27.282] Stopping MConnection service                 module=server impl=MConn{pipe}
--- FAIL: TestMConnectionChannelOverflow (0.00s)
FAIL
exit status 1
FAIL	github.com/tendermint/tendermint/p2p/conn	0.418s
```

--- a/p2p/conn/connection.go
+++ b/p2p/conn/connection.go
@@ -625,8 +625,9 @@ FOR_LOOP:
 				// never block
 			}
 		case *tmp2p.Packet_PacketMsg:
-			channel, ok := c.channelsIdx[byte(pkt.PacketMsg.ChannelID)]
-			if !ok || channel == nil {
+			channelID := byte(pkt.PacketMsg.ChannelID)
+			channel, ok := c.channelsIdx[channelID]
+			if pkt.PacketMsg.ChannelID < 0 || pkt.PacketMsg.ChannelID > math.MaxUint8 || !ok || channel == nil {
 				err := fmt.Errorf("unknown channel %X", pkt.PacketMsg.ChannelID)
 				c.Logger.Debug("Connection failed @ recvRoutine", "conn", c, "err", err)
 				c.stopForError(err)
@@ -642,9 +643,9 @@ FOR_LOOP:
 				break FOR_LOOP
 			}
 			if msgBytes != nil {
-				c.Logger.Debug("Received bytes", "chID", pkt.PacketMsg.ChannelID, "msgBytes", msgBytes)
+				c.Logger.Debug("Received bytes", "chID", channelID, "msgBytes", msgBytes)
 				// NOTE: This means the reactor.Receive runs in the same thread as the p2p recv routine
-				c.onReceive(byte(pkt.PacketMsg.ChannelID), msgBytes)
+				c.onReceive(channelID, msgBytes)
 			}
 		default:
 			err := fmt.Errorf("unknown message type %v", reflect.TypeOf(packet))

--- a/p2p/conn/connection_test.go
+++ b/p2p/conn/connection_test.go
@@ -553,6 +553,36 @@ func TestConnVectors(t *testing.T) {
 	}
 }
 
+func TestMConnectionChannelOverflow(t *testing.T) {
+	chOnErr := make(chan struct{})
+	chOnRcv := make(chan struct{})
+
+	mconnClient, mconnServer := newClientAndServerConnsForReadErrors(t, chOnErr)
+	t.Cleanup(stopAll(t, mconnClient, mconnServer))
+
+	mconnServer.onReceive = func(chID byte, msgBytes []byte) {
+		chOnRcv <- struct{}{}
+	}
+
+	client := mconnClient.conn
+	protoWriter := protoio.NewDelimitedWriter(client)
+
+	var packet = tmp2p.PacketMsg{
+		ChannelID: 0x01,
+		EOF:       true,
+		Data:      []byte(`42`),
+	}
+	_, err := protoWriter.WriteMsg(mustWrapPacket(&packet))
+	require.NoError(t, err)
+	assert.True(t, expectSend(chOnRcv))
+
+	packet.ChannelID = int32(1025)
+	_, err = protoWriter.WriteMsg(mustWrapPacket(&packet))
+	require.NoError(t, err)
+	assert.False(t, expectSend(chOnRcv))
+
+}
+
 type stopper interface {
 	Stop() error
 }

